docs/README.md

```diff
@@ -714,6 +714,246 @@ The `:=`/`:=@` syntax is JSON-specific. You can switch your request to `--form`
 Other JSON types, however, are not allowed with `--form` or `--multipart`.
 
+### Nested JSON fields
+
+For creating nested JSON structures, you can simply declare the path for the object's new destination
+and HTTPie will interpret it according to the [JSON form](https://www.w3.org/TR/html-json-forms/)
+notation and create your object. It works directly with the existing data field (`=`) and raw JSON
+field (`:=`) operators.
+
+#### Path Declaration
+
+A simple path can be a shallow key;
+
+```bash
+$ http --offline --print=B pie.dev/post \
+  type=success
+```
+
+```json
+{
+    "type": "success"
+}
+```
+
+As well as a nested one,
+
+```bash
+$ http --offline --print=B pie.dev/post \
+  result[type]=success
+```
+
+```json
+{
+    "result": {"type": "success"}
+}
+```
+
+Or even multiple levels of nesting.
+
+```bash
+$ http --offline --print=B pie.dev/post \
+  result[status][type]=ok
+```
+
+```json
+{
+    "result": {
+        "status": {
+            "type": "ok"
+        }
+    }
+}
+```
+
+The declaration also supports creating arrays; which can be either done by simply
+assigning the same path multiple times
+
+```bash
+$ http --offline --print=B pie.dev/post \
+  ids:=1 ids:=2
+```
+
+```json
+{
+    "ids": [
+        1,
+        2
+    ]
+}
+```
+
+Or using the append suffix `[]`, which would create an array and append the items to the
+end of it.
+
+```bash
+$ http --offline --print=B pie.dev/post \
+  ids[]:=1
+```
+
+```json
+{
+    "ids": [
+        1,
+        2
+    ]
+}
+```
+
+You can also use indexes to set items on an array,
+
+```bash
+$ http --offline --print=B pie.dev/post \
+  items[0]=terminal items[1]=desktop
+```
+
+```json
+{
+    "items": [
+        "terminal",
+        "desktop"
+    ]
+}
+```
+
+If you don't set value for the indexes between, then those will be nullified.
+
+```bash
+$ http --offline --print=B pie.dev/post \
+  items[1]=terminal items[3]=desktop
+```
+
+```json
+{
+    "items": [
+        null,
+        "terminal",
+        null,
+        "desktop"
+    ]
+}
+```
+
+It is permitted to mix index-access with append actions (`[]`), but be aware that appends will not fill
+the voids but instead they will append after the last item.
+
+```bash
+$ http --offline --print=B pie.dev/post \
+  items[1]=terminal items[3]=desktop items[]=web
+```
+
+```json
+{
+    "items": [
+        null,
+        "terminal",
+        null,
+        "desktop",
+        "web"
+    ]
+}
+```
+
+If you need to send a top-level list (without any object that is encapsulating it), use the append operator (`[]`) without
+any keys.
+
+```bash
+$ http --offline --print=B pie.dev/post \
+  []:=1 []:=2 []:=3
+```
+
+```json
+[
+    1,
+    2,
+    3
+]
+```
+
+Here is a slightly unified example
+
+```bash
+$ http --offline --print=B pie.dev/post name=python version:=3 \
+  date[year]:=2021 date[month]=December \
+  systems=Linux systems=Mac systems=Windows \
```

Comment:
This actually removes the possibility to overwrite fields:

* `$ http --offline pie.dev a=1 a=2`
  * Old  → `{"a": 2}`
  * New → `{"a": [1, 2]}`

🤔

At some point I considered adding something like `--no-nested-json` to ease a bit the backward compatibility breakage. I.e., if my requests start breaking, I could just turn the nesting off without having to think too much.  What do you think?

Reply:
> At some point I considered adding something like --no-nested-json to ease a bit the backward compatibility breakage

I personally don't really prefer feature togglers, since these features are now part of the essential HTTPie DSL and it might just confuse people. Since depending on a CLI flag, the whole language processing changes. I'd prefer removing this behaviour, and only supporting array creation through `[]` and `[<index>]` over `--no-nested-json` flag. (If there aren't any more pitfalls).


The only bad side about this removal is that it would make us diverge from JSON form notation a bit more, which I think is acceptable considering the notation is not designed for CLI apps to use.

Reply:
Though now thinking a bit more, `=` already creates a list on `--form` mode. The same goes for `:` too. So I don't really know if overwriting should be incoherently supported in only JSON.

Reply:
> The same goes for `:` too. So I don't really know if overwriting should be incoherently supported in only JSON.

For headers, the difference is that we have an [un-setting syntax](https://httpie.io/docs#empty-headers-and-header-un-setting) (just `Header:`). The idea is to allow you to overwrite a header specified in another way, e.g., in a session or the config file. 

For data fields, though, we can more strongly assume that the full payload is specified for each invocation, so the ability to overwrite previously specified fields is probably less important.

Reply:
> For headers, the difference is that we have an un-setting syntax (just Header:). The idea is to allow you to overwrite a header specified in another way, e.g., in a session or the config file.

Makes sense. Forgot about that.

> For data fields, though, we can more strongly assume that the full payload is specified for each invocation, so the ability to overwrite previously specified fields is probably less important.

So, how should we proceed? Leave it as is or should we consider other options (`--no-nested-json`/`not creating an array on path repetition`)?

Reply:
I’d say leave it as it is (if you agree with my reasoning above).

Reply:
I do.